### PR TITLE
backend/dvarstats: Use a D array instead of a pointer + length

### PR DESCRIPTION
**Pull request by Geod24**

### CODE REVIEW

File: src/dmd/backend/dvarstats.d [hunk 1]
```diff
@@ -232,11 +230,8 @@ private symtab_t* calcLexicalScope(symtab_t* symtab) return
     sortLineOffsets();
 
     // precalc the lexical blocks to emit so that identically named symbols don't overlap
-    if (cntAllocLifeTimes < dupcnt)
-    {
-        lifeTimes = cast(LifeTime*) util_realloc(lifeTimes, dupcnt, (LifeTime).sizeof);
-        cntAllocLifeTimes = dupcnt;
-    }
+    if (lifeTimes.length < dupcnt)
+        lifeTimes = (cast(LifeTime*)util_realloc(lifeTimes.ptr, dupcnt, (LifeTime).sizeof))[0 .. dupcnt];
```
Comment (MoonlightSentinel) [inline src/dmd/backend/dvarstats.d:234]:
```suggestion
        lifeTimes = (cast(LifeTime*) util_realloc(lifeTimes.ptr, dupcnt, (LifeTime).sizeof))[0 .. dupcnt];
```

File: src/dmd/backend/dvarstats.d [hunk 2]
```diff
@@ -245,7 +240,7 @@ private symtab_t* calcLexicalScope(symtab_t* symtab) return
         lifeTimes[si].offDestroy = cast(int)getLineOffset(lifeTimes[si].sym.lnoscopeend);
     }
     cntUsedLifeTimes = dupcnt;
-    qsort(lifeTimes, dupcnt, (LifeTime).sizeof, &cmpLifeTime);
+    qsort(lifeTimes.ptr, dupcnt, (lifeTimes[0]).sizeof, &cmpLifeTime);
```
Comment (MoonlightSentinel) [inline src/dmd/backend/dvarstats.d:243]: Why `lifeTimes[0]` instead of `lifeTime`?

File: src/dmd/backend/dvarstats.d [hunk 3]
```diff
@@ -423,10 +418,10 @@ public void recordLineOffset(Srcpos src, targ_size_t off)
         return;
     }
     // don't care for lineOffsets being ordered now, that is taken care of later (calcLexicalScope)
-    if (cntUsedLineOffsets >= cntAllocLineOffsets)
+    if (lineOffsets.length <= cntUsedLineOffsets)
     {
-        cntAllocLineOffsets = 2 * cntUsedLineOffsets + 16;
-        lineOffsets = cast(LineOffset*) util_realloc(lineOffsets, cntAllocLineOffsets, (*lineOffsets).sizeof);
+        const newSize = 2 * cntUsedLineOffsets + 16;
+        lineOffsets = (cast(LineOffset*)util_realloc(lineOffsets.ptr, newSize, (lineOffsets[0]).sizeof))[0 .. newSize];
```
Comment (MoonlightSentinel) [inline src/dmd/backend/dvarstats.d:424]: Ditto

**Review by MoonlightSentinel — APPROVED** [review]

File: src/dmd/backend/dvarstats.d (hunk 2 above)
Comment (Geod24) [reply src/dmd/backend/dvarstats.d:243]: Consistency with the other line where the type of the element is used instead of the type.

Comment (Geod24) [reply src/dmd/backend/dvarstats.d:243]: Although I could go with either, as long as we stay consistent.

Comment (MoonlightSentinel) [reply src/dmd/backend/dvarstats.d:243]: Fair enough

File: src/dmd/backend/dvarstats.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/backend/dvarstats.d:234]: Done

File: src/dmd/backend/dvarstats.d [hunk 4]
```diff
@@ -121,7 +119,7 @@ private extern (C) static int cmpLifeTime(scope const void* p1, scope const void
 }
 
 // a parent scope contains the creation offset of the child scope
-private static SYMIDX isParentScope(LifeTime* lifetimes, SYMIDX parent, SYMIDX si)
```
Comment (MoonlightSentinel) [inline src/dmd/backend/dvarstats.d:124]: Why did you remove the `static`?

Comment (Geod24) [reply src/dmd/backend/dvarstats.d:124]: Oh,  thought this was at module level because it is not indented....

Comment (Geod24) [reply src/dmd/backend/dvarstats.d:124]: Reverted

**Review by MoonlightSentinel — APPROVED** [review]

**Review by thewilsonator — APPROVED** [review]
